### PR TITLE
Use ClientOptions to create Pipeline

### PR DESCRIPTION
**Pull request by heaths**

### CODE REVIEW

File: sdk/core/src/options.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    pub(crate) telemetry: TelemetryOptions,
+
+    /// Transport options.
+    pub(crate) transport: TransportOptions,
+}
+
+impl ClientOptions {
+    setters! {
+        per_call_policies: Vec<Arc<dyn Policy>> => per_call_policies,
+        per_retry_policies: Vec<Arc<dyn Policy>> => per_retry_policies,
```
Comment (rylev) [inline sdk/core/src/options.rs:29]: We may want to expose a setter for these that returns an `&mut Vec<Arc<dyn Policy>>` since it's probably more likely that users want to add policies rather than outright replace the entire collection.

File: sdk/core/src/options.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
+use crate::policies::{ExponentialRetryPolicy, FixedRetryPolicy, NoRetryPolicy, Policy};
+use crate::{new_http_client, HttpClient};
+use std::sync::Arc;
+use std::time::Duration;
+
+/// Client options allow customization of policies, retry options, and more.
```
Comment (rylev) [inline sdk/core/src/options.rs:6]: Can we add a `# Example` section that shows the usage of this type off in code?

**Review by rylev — CHANGES_REQUESTED** [review]
Looking good! I'd like to see this used in the examples folder in Cosmos (at least once) just so we can see some of this in action.

File: sdk/core/src/macros.rs [hunk 3]
```diff
@@ -26,6 +26,7 @@
 #[macro_export]
 macro_rules! setters {
     (@single $name:ident : $typ:ty => $transform:expr) => {
+        // TODO: Declare with_$name when https://github.com/rust-lang/rust/issues/29599 is fixed.
```
[the remark this answers is not in the document]
Comment (heaths) [reply sdk/core/src/macros.rs:29]: The code as-is doesn't actually agree with the comment about it, that says using `setters!()` is equivalent to generating `with_$name` methods. That's not currently possible without taking a dependency on 1 of a couple crates out there that provide this, despite nightly toolset support since ~2016 (still not in stable). I'm capturing the tracking work to make the macro comment actually true, or we need to take a dependency (or duplicate) on those crates, or fix the comment so it doesn't say `with_$name` methods are generated, but that doesn't seem very idiomatic.

File: sdk/core/src/options.rs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        RetryMode::Exponential
+    }
+}
+
+/// The set of options that can be specified to influence how retry attempts are made,
+/// and a failure is eligible to be retried.
+#[derive(Clone, Debug)]
+pub struct RetryOptions {
+    /// The algorithm to use for calculating retry delays.
+    pub(crate) mode: RetryMode,
```
[the remark this answers is not in the document]
Comment (heaths) [reply sdk/core/src/options.rs:63]: I need to access these in scopes that are children of a sibling scope.

File: sdk/core/src/options.rs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+impl TransportOptions {
+    pub fn new(http_client: Arc<dyn HttpClient>) -> Self {
+        Self { http_client }
+    }
+}
+
+impl Default for TransportOptions {
+    /// Creates an instance of the ```TransportOptions``` using the enabled HTTP client.
```
[the remark this answers is not in the document]
Comment (heaths) [reply sdk/core/src/options.rs:147]: I know for markdown in general, but doesn't rustdoc ensure that code - even tokens - in triplebackticks is valid, or is that only for codeblocks and not inline code?

PS: There's a lot of other code under services/ that does this. I write a ton of markdown and only use a single backtick for inline code normally.

[the remark this answers is not in the document]
Comment (heaths) [reply sdk/core/src/options.rs:147]: Just tried making a syntax error in that case and `cargo test` nor `cargo doc` cares so I'll use proper inline code.

File: sdk/core/src/policies/transport.rs [hunk 6]
```diff
@@ -1,28 +1,20 @@
...
 #[derive(Debug, Clone)]
 pub struct TransportPolicy {
-    options: TransportOptions,
+    http_client: Arc<dyn HttpClient>,
 }
 
 impl TransportPolicy {
-    pub fn new(options: TransportOptions) -> Self {
-        Self { options }
+    pub fn new(options: &TransportOptions) -> Self {
```
[the remark this answers is not in the document]
Comment (heaths) [reply sdk/core/src/policies/transport.rs:14]: 1. I did so in TelemetryPolicy because copying or taking ownership of the options seems pointless for read-only operations.
2. Arc doesn't implement the `Copy` trait, as rustc was complaining hence the explicit `clone` call as well. Taking this by-ref not only saves the copy but resolve that issue as well.

File: sdk/core/src/policies/retry_policies/exponential_retry.rs [hunk 7]
```diff
@@ -18,17 +18,15 @@ pub struct ExponentialRetryPolicy {
...
-        Self {
-            delay: Duration::from_secs(3),
-            max_retries: 3,
-            max_delay: Duration::from_secs(30),
+impl ExponentialRetryPolicy {
+    pub(crate) fn new(delay: Duration, max_retries: u32, max_delay: Duration) -> Self {
+        ExponentialRetryPolicy {
+            delay: delay,
+            max_retries: max_retries,
+            max_delay: max_delay,
```
[the remark this answers is not in the document]
Comment (heaths) [reply sdk/core/src/policies/retry_policies/exponential_retry.rs:26]: Thanks. I forgot about that. Odd that `cargo fmt` doesn't do that.

File: sdk/core/src/macros.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (heaths) [reply sdk/core/src/macros.rs:29]: Ah, I forgot I fixed the name. Still, excluding "with_" as a function prefix doesn't seem very idiomatic. I think we should change this to actually be `with_$name` before we GA if that's truly what's idiomatic. I'll open a tracking issue of our own.

[the remark this answers is not in the document]
Comment (heaths) [reply sdk/core/src/macros.rs:29]: Opened https://github.com/Azure/azure-sdk-for-rust/issues/292. We do track issues this way in our other SDK repos as well.

**Comment by heaths** [comment]
Oh, and I forgot to add some more examples like you mentioned in your overarching comment. I'll do that this weekend before merging if you feel appropriate to change your vote.

**Comment by heaths** [comment]
I opened https://github.com/Azure/azure-sdk-for-rust/issues/294 to track moving (exposing) request methods from the pipeline itself, like with our other SDKs, to abstract away the requests from the wire protocol.

File: sdk/core/src/options.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ctaggart) [reply sdk/core/src/options.rs:84]: I am not sure I understand why. I checkout out this branch, removed the `pub(crate)` and it compiles.

**Review by ctaggart — APPROVED** [review]
Pipeline is looking better. I especially like the example in doc comments.

File: sdk/core/src/options.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (heaths) [reply sdk/core/src/options.rs:84]: I'll try it. The docs said (paraphrasing - or perhaps I misunderstood entirely) that `pub(crate)` was necessary to access members of sibling mods, or children of sibling mods.

[the remark this answers is not in the document]
Comment (heaths) [reply sdk/core/src/options.rs]: I was able to remove them from some of the members but not all. Started with removing them all and added them back in as needed.

**Comment by heaths** [comment]
I opened #300 to further the discussion on how we want the guidelines to end up in this regard.

File: sdk/core/src/options.rs [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    /// Telemetry options.
+    pub(crate) telemetry: TelemetryOptions,
+
+    /// Transport options.
+    pub(crate) transport: TransportOptions,
+}
+
+impl ClientOptions {
+    /// A mutable reference to per-call policies.
+    pub fn mut_per_call_policies(&mut self) -> &mut Vec<Arc<dyn Policy>> {
```
Comment (rylev) [inline sdk/core/src/options.rs:39]: The idiomatic form is `per_call_policies_mut`

File: sdk/core/src/options.rs [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        &mut self.per_call_policies
+    }
+
+    /// A mutable reference to per-retry policies.
+    pub fn mut_per_retry_policies(&mut self) -> &mut Vec<Arc<dyn Policy>> {
```
Comment (rylev) [inline sdk/core/src/options.rs:44]: Same as above

File: sdk/core/src/options.rs (hunk 1 above)
Comment (rylev) [reply sdk/core/src/options.rs:50]: Do we need to keep these with the new `mut` setters?

File: sdk/core/src/pipeline.rs [hunk 10]
```diff
@@ -51,17 +52,33 @@ impl Pipeline {
...
+
+        Self {
+            http_client: options.transport.http_client.clone(),
+            pipeline,
+        }
+    }
+
+    /// Gets the `HttpClient` used by the pipeline.
+    pub fn http_client(&self) -> &dyn HttpClient {
+        // TODO: Request methods should be defined directly on the pipeline instead of exposing the HttpClient.
```
Comment (rylev) [inline sdk/core/src/pipeline.rs:80]: Can you create an issue for this?

File: sdk/core/src/policies/retry_policies/fixed_retry.rs [hunk 11]
```diff
@@ -4,30 +4,28 @@ use chrono::{DateTime, Local};
...
 
-impl Default for LinearRetryPolicy {
-    fn default() -> Self {
-        Self {
-            delay: Duration::from_secs(3),
-            max_retries: 3,
-            max_delay: Duration::from_secs(10),
+impl FixedRetryPolicy {
+    pub(crate) fn new(delay: Duration, max_retries: u32, max_delay: Duration) -> Self {
+        FixedRetryPolicy {
```
Comment (rylev) [inline sdk/core/src/policies/retry_policies/fixed_retry.rs:22]: Why the change away from `Self`?

**Review by rylev — APPROVED** [review]
Looks good. I had some additional questions and nits, but I'd like to see this merged ASAP

File: sdk/core/src/pipeline.rs (hunk 10 above)
Comment (rylev) [reply sdk/core/src/pipeline.rs:80]: Ah I see that there's this: https://github.com/Azure/azure-sdk-for-rust/issues/294. Might be worth in the future having all todos link to an issue
